runtime/lua/vim/shared.lua

```diff
@@ -548,6 +548,45 @@ function vim.tbl_islist(t)
   end
 end
 
+--- Tests if a Lua table can be treated as a list (a table indexed by consecutive integers starting from 1).
+---
+--- Empty table `{}` is assumed to be an list, unless it was created by
+--- |vim.empty_dict()| or returned as a dict-like |API| or Vimscript result,
+--- for example from |rpcrequest()| or |vim.fn|.
+---
+---@param t table
+---@return boolean `true` if list-like table, else `false`.
+function vim.tbl_islist(t)
+  if type(t) ~= 'table' then
+    return false
+  end
+
+  local count = 0
+  local max_idx = 0
+
+  for k, _ in pairs(t) do
+    if type(k) == 'number' and k > 0 then
+      count = count + 1
+      if max_idx < k then
+        max_idx = k
+      end
+    else
+      return false
+    end
+  end
+
+  if count > 0 then
+    return count == max_idx
+  else
+    -- TODO(bfredl): in the future, we will always be inside nvim
+    -- then this check can be deleted.
+    if vim._empty_dict_mt == nil then
+      return nil
+    end
```

Comment:
@clason Should I still keep this TODO?

Reply:
That's right, I was meaning to ask @bfredl whether we are already in the future now!

Reply:
I haven't heard back, and I'll leave this one for now.